cpp/include/rmm/mr/cuda_async_pinned_memory_resource.hpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#pragma once
+
+#include <rmm/cuda_device.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/detail/error.hpp>
+#include <rmm/detail/export.hpp>
+#include <rmm/detail/runtime_capabilities.hpp>
+#include <rmm/detail/thrust_namespace.h>
+#include <rmm/mr/cuda_async_view_memory_resource.hpp>
+#include <rmm/mr/device_memory_resource.hpp>
+
+#include <cuda/std/type_traits>
+#include <cuda_runtime_api.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <optional>
+
+namespace RMM_NAMESPACE {
+namespace mr {
+/**
+ * @addtogroup memory_resources
+ * @{
+ * @file
+ */
+
+/**
+ * @brief `device_memory_resource` derived class that uses
+ * `cudaMallocFromPoolAsync`/`cudaFreeFromPoolAsync` with a pinned memory pool
+ * for allocation/deallocation.
+ */
+class cuda_async_pinned_memory_resource final : public device_memory_resource {
+ public:
+  /**
+   * @brief Constructs a cuda_async_pinned_memory_resource with a pinned memory pool for
+   * the current device.
+   *
+   * On CUDA 12.6-12.x, creates a new pinned memory pool using cudaMemPoolCreate.
+   * On CUDA 13.0+, uses the default pinned memory pool via cudaMemGetDefaultMemPool.
+   *
+   * Pool properties such as the release threshold are not modified.
+   *
+   * @throws rmm::logic_error if the CUDA build version is less than 12.6
+   * @throws rmm::logic_error if the CUDA runtime version does not support pinned memory pools
+   * (requires CUDA 12.6 or higher)
+   */
+  cuda_async_pinned_memory_resource()
+  {
+#if !defined(CUDA_VERSION) || CUDA_VERSION < RMM_MIN_ASYNC_PINNED_ALLOC_CUDA_VERSION
+    RMM_FAIL(
+      "cuda_async_pinned_memory_resource requires CUDA 12.6 or higher. "
+      "This build was compiled with an older CUDA version.");
+#else
+    // Check if pinned memory pools are supported at runtime
+    RMM_EXPECTS(rmm::detail::runtime_async_pinned_alloc::is_supported(),
+                "cuda_async_pinned_memory_resource requires CUDA 12.6 or higher runtime");
+
+    pool_handle_.reset(new cudaMemPool_t{});
+
+#if CUDA_VERSION >= 13000
+    // CUDA 13.0+: Use the default pinned memory pool (no cleanup needed)
+    cudaMemLocation location{.type = cudaMemLocationTypeDevice,
+                             .id   = rmm::get_current_cuda_device().value()};
+    RMM_CUDA_TRY(
+      cudaMemGetDefaultMemPool(pool_handle_.get(), &location, cudaMemAllocationTypePinned));
+#else
+    // CUDA 12.6-12.x: Create a new pinned memory pool (needs cleanup)
+    cudaMemPoolProps pool_props{};
+    pool_props.allocType     = cudaMemAllocationTypePinned;
+    pool_props.location.type = cudaMemLocationTypeDevice;
```

Comment:
This is making a location as `DEVICE`. Is this correct? 
In CCCL pinned mem pool, its marked as host/ host_numa 
https://github.com/NVIDIA/cccl/blob/main/libcudacxx/include/cuda/__memory_resource/pinned_memory_pool.h#L113-L154